compiler-core/src/type_/expression.rs

```diff
@@ -2410,7 +2487,19 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
             (_, value) => self.infer(value),
         }?;
 
-        unify(typ, value.type_()).map_err(|e| convert_unify_error(e, value.location()))?;
+        unify(typ, value.type_()).map_err(|e| match kind {
+            ArgumentKind::UseCallback {
+                function_location,
+                assignments_location,
+                last_statement_location,
+            } => e.into_use_unify_error(
+                function_location,
+                assignments_location,
+                last_statement_location,
+                value.location(),
+            ),
+            ArgumentKind::Regular => convert_unify_error(e, value.location()),
+        })?;
```

Comment:
Could you extract this into a method please? I'm finding it a bit hard to see the logic as the majority of the code is type conversion boilerplate.